### PR TITLE
Ropes Issue #12 - Authentication - Part 1

### PR DESCRIPTION
**Pull request by zldavis**
Users can sign up.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -22,7 +22,7 @@ gem 'jbuilder', '~> 2.5'
 # Use Redis adapter to run Action Cable in production
```
Comment (bboe) [inline Gemfile]: With regard to your commit messages, e.g., "user model", I feel there is some to be desired. For instance, without context I don't know if you're touching the user model, adding the user model, or deleting the user model. Writing "Add user model" is certainly sufficient to state that is done.

There's more info here on writing useful commit messages, though I'll admit I sometimes save the why for the PR rather than the individual commit:
https://chris.beams.io/posts/git-commit/

File: test/models/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  def setup
+    @user = User.new(name: 'DHH', email: 'dhh@mailinator.com', password: 'basecamp',
+                     password_confirmation: 'basecamp')
+  end
+
+  test 'should be valid' do
+    assert @user.valid?
```
Comment (bboe) [inline test/models/user_test.rb]: I refer to these naked asserts. Compare the output when this assertion fails to the output provided by `assert_predicate @user, :valid?`. There's more information here:

https://appfolio.slack.com/archives/C02A9D6U3/p1496265219443302

File: test/models/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  test 'name is required' do
+    @user.name = nil
+    refute @user.valid?
```
Comment (bboe) [inline test/models/user_test.rb]: There's a side effect when something is not valid. That is that rails sets an error message. Can you make an assertion about those error messages that expect only the error message that you expect it to have? In that way you can guarantee it doesn't fail for a different reason than the one your expecting.

File: test/models/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  test 'email is required' do
+    @user.email = nil
+    refute @user.valid?
+
+    @user.email = ' '
+    refute @user.valid?
+  end
+
+  test 'email must be valid email format' do
+    %w(foo foo@bar bar.com).each do |email|
```
Comment (bboe) [inline test/models/user_test.rb]: This is a tricky one. There are a ton of inclusion and exclusion states in this regex: `/\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i`. I like to make the argument that at least one condition is tested for every inclusion/exclusion state, and all the boundary cases are tested, which in this case is the 0, 1, and 2 cases to represent the `+` in the regex.

The reason for the pedantic tests are to ensure if someone modifies the regex that they change only what they intend. With that said, perhaps there is a gem you can rely on that does some level of email validation and you can "assume" that said library is accurate enough to avoid skipping the pedantic tests.

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  test 'password is required' do
+    @user.password = @user.password_confirmation = nil
+    refute @user.valid?
+
+    @user.password = @user.password_confirmation = ' ' * 6
+    refute @user.valid?
+  end
+
+  test 'password must be at least 6 characters' do
```
Comment (bboe) [inline test/models/user_test.rb]: Great test.

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  before_save { self.email = email.downcase }
+
+  validates :name, presence: true, length: { maximum: 50 }
```
Comment (bboe) [inline app/models/user.rb]: Can you test the boundary conditions for having a 50 character name?

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 }, format: { with: VALID_EMAIL_REGEX },
```
Comment (bboe) [inline app/models/user.rb]: Can you test the 255 email length boundary condition?

File: test/models/user_test.rb (hunk 2 above)
Comment (zldavis) [inline test/models/user_test.rb]: thanks, makes sense.

File: Gemfile (hunk 1 above)
Comment (zldavis) [inline Gemfile]: That was a good, thanks for sharing.

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+
+    if @user.save
+      flash[:success] = "Welcome to Image Sharer #{@user.name}!"
+      redirect_to root_path
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Just as a heads up, you can set a flash message directly as part of `redirect_to`. I believe the syntax would be:

    redirect_to root_path, flash: { success: 'Welcome ...' }

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    else
+      render :new
```
Comment (bboe) [inline app/controllers/users_controller.rb]: What HTTP response code does this result in?

File: test/controllers/users_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'new renders view with form' do
+    get new_user_path
+    assert_response :ok
+    assert_select 'form[method="post"][action="/users"]'
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Awesome test.

File: test/controllers/users_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    assert_redirected_to root_path
+    assert_equal 'Welcome to Image Sharer DHH!', flash[:success]
+  end
+
+  test 'emails must be unique' do
+    User.create!(name: 'DHH', email: 'boss@rails.org', password: 'basecamp', password_confirmation: 'basecamp')
+
+    post users_path, params: {
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: What's the response code?

File: test/controllers/users_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'new renders view with form' do
+    get new_user_path
+    assert_response :ok
+    assert_select 'form[method="post"][action="/users"]'
+  end
+
+  test 'user can sign up' do
+    post users_path, params: {
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: How can we assert that the database side effect occurred?

File: test/controllers/users_controller_test.rb (hunk 11 above)
Comment (bboe) [inline test/controllers/users_controller_test.rb]: In this case we expect there is no database side effect. How can we ensure that doesn't occur?

File: test/controllers/users_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      user: { name: 'DHH', email: 'da_boss@rails.org', password: 'basecamp', password_confirmation: 'basecamp' }
+    }
+
+    assert_redirected_to root_path
+    assert_equal 'Welcome to Image Sharer DHH!', flash[:success]
+  end
+
+  test 'emails must be unique' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Consider making this test more generic. E.g., 'user fails to sign up'. Since these two distinct cases appear to both test validations, of which we only really need one at the controller level.

The exception would be if there were different failure cases in the controller, but in this case there is only a single path to failure.
